cl/persistence/beacon_indexes/sql_indexer.go

```diff
@@ -0,0 +1,138 @@
+package beacon_indexes
+
+import (
+	"database/sql"
+	"fmt"
+
+	libcommon "github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/erigon/cl/cltypes"
+	_ "github.com/mattn/go-sqlite3"
+)
+
+// Define all queries templates
+var beaconIndiciesCreateTableQuery = `CREATE TABLE beacon_indicies (
+    Slot INTEGER NOT NULL,
+	BeaconBlockRoot BLOB NOT NULL CHECK(length(BeaconBlockRoot) = 32), -- Ensure it's 32 bytes
+	StateRoot BLOB NOT NULL CHECK(length(BeaconBlockRoot) = 32),
+	ParentBlockRoot BLOB NOT NULL CHECK(length(BeaconBlockRoot) = 32),
+    Canonical INTEGER NOT NULL DEFAULT 0, -- 0 for false, 1 for true
+    PRIMARY KEY (Slot, BeaconBlockRoot)  -- Composite key ensuring unique combination of Slot and BeaconBlockRoot
+);`
+
+var beaconIndiciesUniqueIndiciesQuery = `
+	CREATE UNIQUE INDEX idx_unique_canonical 
+	ON beacon_indicies (Slot) 
+	WHERE Canonical = 1;`
+
+var beaconReadBlockSlotByBlockRoot = "SELECT Slot FROM beacon_indicies WHERE BeaconBlockRoot = ?"
+
+var beaconReadParentRootByBlockRoot = "SELECT ParentBlockRoot FROM beacon_indicies WHERE BeaconBlockRoot = ?"
+
+var beaconReadCanonicalBlockRootBySlot = `SELECT BeaconBlockRoot FROM beacon_indicies 
+	WHERE Slot = ? AND Canonical = 1`
+
+var markRootCanonicalQuery = `UPDATE beacon_indicies 
+	SET Canonical = 1 
+	WHERE Slot = ? AND BeaconBlockRoot = ?`
+
+var resetOtherRootsCanonicalQuery = `UPDATE beacon_indicies 
+	SET Canonical = 0 
+	WHERE Slot = ? AND BeaconBlockRoot != ?`
+
+var insertOrUpdateBlockRoot = `
+	INSERT INTO beacon_indicies (Slot, BeaconBlockRoot, StateRoot, ParentBlockRoot, Canonical) 
+	VALUES (?, ?, ?, ?, 0);
+`
+
+type SqliteBeaconIndexer struct {
+	db *sql.DB
+}
```

Comment:
technically, this is bad practice.

functions that use the database should have the following signature: 

```
function SomeFunc(ctx context.Context, db *sql.DB, <receiver>, <more args>) error
```

Reply:
you should put the DB in your larger struct, and then use it when calling your "sql" functions.

putting your sql calls behind a struct is bad.  

to honor the interface, create a struct that has the db in it, and then pass db with external context to create struct that honor interrface